grpc/tests/message_builder_test.cpp

```diff
@@ -0,0 +1,148 @@
+#include <iostream>
+#include "flatbuffers/grpc.h"
+#include "monster_test_generated.h"
+
+using namespace MyGame::Example;
+
+const std::string m1_name = "Cyberdemon";
+const Color m1_color = Color_Red;
+const std::string m2_name = "Imp";
+const Color m2_color = Color_Green;
+
+flatbuffers::Offset<Monster> populate1(flatbuffers::FlatBufferBuilder &builder) {
+  auto name_offset = builder.CreateString(m1_name);
+  return CreateMonster(builder, nullptr, 0, 0, name_offset, 0, m1_color);
+}
+
+flatbuffers::Offset<Monster> populate2(flatbuffers::FlatBufferBuilder &builder) {
+  auto name_offset = builder.CreateString(m2_name);
+  return CreateMonster(builder, nullptr, 0, 0, name_offset, 0, m2_color);
+}
+
+bool release_n_verify(flatbuffers::FlatBufferBuilder &fbb, const std::string &expected_name, Color color) {
+  flatbuffers::DetachedBuffer buf = fbb.Release();
+  const Monster *monster = flatbuffers::GetRoot<Monster>(buf.data());
+  return (monster->name()->str() == expected_name) && (monster->color() == color);
+}
+
+bool release_n_verify(flatbuffers::grpc::MessageBuilder &mbb, const std::string &expected_name, Color color) {
+  flatbuffers::grpc::Message<Monster> msg = mbb.ReleaseMessage<Monster>();
+  const Monster *monster = msg.GetRoot();
+  return (monster->name()->str() == expected_name) && (monster->color() == color);
+}
+
+struct TestHeapMessageBuilder : public flatbuffers::FlatBufferBuilder {
+  TestHeapMessageBuilder()
+    : flatbuffers::FlatBufferBuilder(2048, new flatbuffers::DefaultAllocator(), true) {}
```

Comment:
Why are you allocating a `DefaultAllocator` explicitly?

Reply:
There are three cases
1. `FlatBufferBuilder` with a null allocator--`FlatBufferBuilder` defaults to `DefaultAllocator` internally.
2. `MessageBuilder` with a funky allocator that's a member--`FlatBufferBuilder` uses the passed allocator in non-owned manner.
3. `TestHeapMessageBuilder` with a custom allocator (owned) that's neither option 1 nor 2. I used `DefaultAllocator` as it was handy. It could be anything that `is-a` `flatbuffers::Allocator`.

May be this is clearer.
```cpp
struct OwnedAllocator : public flatbuffers::DefaultAllocator {};

struct TestHeapMessageBuilder : public flatbuffers::FlatBufferBuilder {
  TestHeapMessageBuilder()
    : flatbuffers::FlatBufferBuilder(2048, new OwnedAllocator(), true) {}
};